AMDGPU/GlobalISel: Import D16 load patterns and add combines for them

petar-avramovic opened the pull request:
Add G_AMDGPU_LOAD_D16 generic instructions and GINodeEquivs for them,
this will import D16 load patterns to global-isel's tablegened
instruction selector.
For newly imported patterns to work add combines for G_AMDGPU_LOAD_D16
in AMDGPURegBankCombiner.

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/load-d16.ll:
```diff
@@ -0,0 +1,412 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc -mtriple=amdgcn -mcpu=gfx1200  -global-isel -new-reg-bank-select < %s | FileCheck --check-prefixes=GFX12 %s
```
```suggestion
; RUN: llc -global-isel -mtriple=amdgcn -mcpu=gfx1200 -new-reg-bank-select < %s | FileCheck --check-prefixes=GFX12 %s
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPURegBankCombiner.cpp:
```diff
@@ -392,6 +401,102 @@ void AMDGPURegBankCombinerImpl::applyCanonicalizeZextShiftAmt(
 
+bool AMDGPURegBankCombinerImpl::matchD16Load(MachineInstr &MI,
+                                             D16MatchInfo &MatchInfo) const {
+  if (!STI.d16PreservesUnusedBits())
```
We do have a way, it's the field called `Predicates` in the combine rule.

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUCombine.td:
```diff
@@ -71,6 +71,14 @@ def int_minmax_to_med3 : GICombineRule<
+
+def d16_load : GICombineRule<
+  (defs root:$bitcast, d16_matchdata:$matchinfo),
+  (match (wip_match_opcode G_BITCAST):$bitcast,
```
Don't introduce more `wip_match_opcode` for insts that aren't variadic
```suggestion
  (match (G_BITCAST $dst, $src):$bitcast,
```

Pierre-vh commented on llvm/lib/Target/AMDGPU/AMDGPUCombine.td:
```diff
@@ -71,6 +71,14 @@ def int_minmax_to_med3 : GICombineRule<
+  (defs root:$bitcast, d16_matchdata:$matchinfo),
+  (match (wip_match_opcode G_BITCAST):$bitcast,
+         [{ return matchD16Load(*${bitcast}, ${matchinfo}); }]),
+  (apply [{ applyD16Load(*${bitcast}, ${matchinfo}); }])>;
```
You could use the `(combine ...` rule to do match + apply in one go, that avoids passing the complex MatchInfo object around

petar-avramovic committed:
AMDGPU/GlobalISel: Import D16 load patterns and add combines for them

Add G_AMDGPU_LOAD_D16 generic instructions and GINodeEquivs for them,
this will import D16 load patterns to global-isel's tablegened
instruction selector.
For newly imported patterns to work add combines for G_AMDGPU_LOAD_D16
in AMDGPURegBankCombiner.
